fe/parameterize_batch.py

```diff
@@ -0,0 +1,523 @@
+    all_du_dls = []
+
+    start = 1e3
+    end = 1.0
+    NT = 500
+    base = np.exp(np.log(end/start)/NT)
+    exps = np.arange(NT)
+    part_one = np.power(base, exps)*start
+    part_two = np.linspace(1.0, 0.3, 1000)
+    part_three = np.linspace(0.3, 0.0, 4000)
+
+    forward_schedule = np.concatenate([part_one, part_two, part_three])
+    backward_schedule = forward_schedule[::-1]
+    lambda_schedule = np.concatenate([forward_schedule, backward_schedule])
+
+    T = lambda_schedule.shape[0]
+    assert T % 2 == 0
+    dt = 0.0015
+    step_sizes = np.ones(T)*dt
+
+    assert T % 2 == 0
+    cas = np.ones(T)*0.93
+
+    host_pdb_file = args.complex_pdb
+    host_pdb = app.PDBFile(host_pdb_file)
+    host_conf = []
+    for x,y,z in host_pdb.positions:
+        host_conf.append([to_md_units(x),to_md_units(y),to_md_units(z)])
+    host_conf = np.array(host_conf)
+    host_name = "complex"
+
+    # set up the system
+    amber_ff = app.ForceField('amber99sbildn.xml', 'amber99_obc.xml')
+    host_system = amber_ff.createSystem(host_pdb.topology,
+        nonbondedMethod=app.NoCutoff,
+        constraints=None,
+        rigidWater=False)
+
+    host_system = openmm_converter.deserialize_system(host_system)
+    num_host_atoms = len(host_system.masses)
+
+    print("num_host_atoms", num_host_atoms)
+
+    for ligand_sdf_file in args.ligand_sdf:
+        for guest_mol in Chem.SDMolSupplier(ligand_sdf_file, removeHs=False):
+            break
+
+    open_ff = forcefield.Forcefield(args.forcefield)
+    nrg_fns = open_ff.parameterize(guest_mol)
+    guest_masses = get_masses(guest_mol)
+    guest_system = system.System(nrg_fns, open_ff.params, open_ff.param_groups, guest_masses)
+
+    combined_system = host_system.merge(guest_system)
+    cbs = -1*np.ones_like(np.array(combined_system.masses))*0.0001
+    lambda_idxs = np.zeros(len(combined_system.masses), dtype=np.int32)
+    lambda_idxs[num_host_atoms:] = -1
+
+    sim = simulation.Simulation(
+        combined_system,
+        step_sizes,
+        cas,
+        cbs,
+        lambda_schedule,
+        lambda_idxs,
+        precision
+    )
+
+    initial_params = sim.system.params
+
+    lr = 5e-4
+    # opt_init, opt_update, get_params = optimizers.adam(lr)
+    opt_init, opt_update, get_params = optimizers.sgd(lr)
+
+    opt_state = opt_init(initial_params)
+
+    num_epochs = 100
+
+    # torsion, charge, gb radii, gb scaling factor, torsion
+    vary_allowed_groups = [(7, 0.5)]*10 + [(14, 0.5)]*20 + [(12, 1e-2)]*20 + [(13, 1e-2)]*20 + [(7, 0.5)]*30
+
+    assert len(vary_allowed_groups) == num_epochs
+
+    ligand_list = []     
+
+    for ligand_sdf_file in args.ligand_sdf:
+        for rd_mol in Chem.SDMolSupplier(ligand_sdf_file):
+            dG_value = rd_mol.GetProp("_dG")
+            ligand_list.append((rd_mol, float(dG_value)))
+
+    ligand_dataset = dataset.Dataset(ligand_list)
+    train, test = ligand_dataset.split(0.7)
+
+    if os.path.exists(args.out_dir+"/loss.txt"):
```

Comment:
what I've been doing is just pipe runs to to a file then grep for LOSS afterwards

Reply:
Yes, that's okay too. I save this just for monitoring the loss more easily.